servers/rendering/renderer_rd/renderer_canvas_render_rd.h

```diff
@@ -398,50 +430,191 @@ class RendererCanvasRenderRD : public RendererCanvasRender {
+			uint32_t hash = hash_murmur3_one_64(texture.get_id());
+			return hash_murmur3_one_32(other, hash);
+		}
 	};
 
-	struct TextureInfo {
-		TextureState state;
-		uint32_t specular_shininess = 0;
-		uint32_t flags = 0;
-		Vector2 texpixel_size;
+	union BatchIndices {
+		struct {
+			uint8_t color;
+			uint8_t normal;
+			uint8_t specular;
+			uint8_t sampler;
+		};
+		/// Little-endian encoding of the indices.
+		/// - bits 0-7  : color
+		/// - bits 8-15 : normal
+		/// - bits 16-23: specular
+		/// - bits 24-31: sampler
+		uint32_t batch_indices;
+
+		BatchIndices() :
+				batch_indices(0x80808080) {}
+
+		_ALWAYS_INLINE_ bool found_textures() const {
+			return (batch_indices & 0x00808080) == 0;
+		}
 
-		RID diffuse;
-		RID normal;
-		RID specular;
-		RID sampler;
+		_ALWAYS_INLINE_ bool found_all() const {
+			return (batch_indices & 0x80808080) == 0;
+		}
 	};
 
+	/// A key used to uniquely identify a set of RIDs that define the BATCH_UNIFORM_SET
+	struct RIDSetKey {
+		RID const *textures = nullptr;
+		uint64_t textures_used = 0;
+		RID const *samplers = nullptr;
+		uint32_t samplers_used = 0;
+		RID instance_data;
+		RID texture_data;
+		/// If _local.is_empty() is false, then the textures and samplers fields point to the _local
+		/// vector, which is used to store the RIDs.
+		Vector<RID> _local;
+
+		RIDSetKey() {
+		}
+
+		RIDSetKey(RID const *p_textures, uint64_t p_textures_used, RID const *p_samplers, uint64_t p_samplers_used, RID p_instance_data, RID p_texture_data) :
+				textures(p_textures),
+				textures_used(p_textures_used),
+				samplers(p_samplers),
+				// The first 12 bits are reserved for the default samplers, which never change,
+				// so we can ignore them.
+				samplers_used(p_samplers_used >> 12),
+				instance_data(p_instance_data),
+				texture_data(p_texture_data) {
+		}
+
+		/// Create a copy of the key, and if the key is not owned, then
+		/// copy the RIDs into the _local vector.
+		RIDSetKey to_owned() {
+			RIDSetKey key = *this;
+			if (_local.is_empty()) {
+				key._local.resize(BATCH_MAX_TEXTURES + BATCH_MAX_DYNAMIC_SAMPLERS);
+				RID *local_ptr = key._local.ptrw();
+				memcpy(local_ptr, textures, sizeof(RID) * BATCH_MAX_TEXTURES);
+				key.textures = local_ptr;
+				local_ptr += BATCH_MAX_TEXTURES;
+
+				if (samplers_used > 0) {
+					memcpy(local_ptr, samplers, sizeof(RID) * BATCH_MAX_DYNAMIC_SAMPLERS);
+					key.samplers = local_ptr;
+				} else {
+					key.samplers = nullptr;
+				}
+			}
+			return key;
+		}
+
+		_ALWAYS_INLINE_ bool operator==(const RIDSetKey &p_val) const {
+			{
+				uint64_t set = textures_used;
+				while (set != 0) {
+					uint32_t index = CTZ64(set);
+					if (textures[index] != p_val.textures[index]) {
+						return false;
+					}
+					// clear the least significant bit that is set
+					// Produces the most efficient code for x86 and ARM architectures
+					// per https://cpp.godbolt.org/z/vjKGPcWjq
+					set &= (set - 1); // clear least significant bit that is set
```

Comment:
Using the new CTZ64 intrinsic. 

Per [Godbolt](https://cpp.godbolt.org/z/vjKGPcWjq), turns out the fastest way to iterate the set bits is using `set &= (set - 1)` vs `set &= ~(1ULL << index)`:

```cpp
while (set != 0) {
  uint32_t index = CTZ64(set);
  // do something ...
  // Option 1
  set &= (set - 1); // clear least significant bit that is set
  // Option 2
  // set &= ~(1ULL << index);
}
```

| option | arch | loop instruction count |
| :--- | :--- | :--- |
| 1 | x64 | 5 |
| 2 | x64 | 8 |
| 1 | arm64 | 6 |
| 2 | arm64 | 7 |